examples/sat/schoening.py

```diff
@@ -1,85 +1,164 @@
+        unsat_clause = inst.clauses[unsat_clauses][0]
+
+        # select a variable that appears in unsatisfied clause, and flip it
+        vars_ = np.argwhere(unsat_clause != 0)
+        var = vars_[step]
+        assignment[var] *= -1
+
+        if inst.evaluate(assignment):
+            return assignment
+
+    return None
+
+
+@attrs.define
+class SchoeningDomain(SamplingDomain[tuple[np.ndarray, np.ndarray]]):
+    """The class SchoeningDomain implements the methods declared for a SamplingDomain. We use this class for search spaces that cannot be stored in memory."""
+
+    n: int
+    n_assignment: int
+    n_walk_steps: int
+
+    @staticmethod
+    def generate_random_assignment(n, rng: np.random.Generator):
+        return rng.integers(2, size=n) * 2 - 1
+
+    @staticmethod
+    def generate_walk_steps(m, rng: np.random.Generator):
+        return rng.integers(3, size=m)
+
+    def get_size(self) -> int:
+        return 2**self.n_assignment * 3**self.n_walk_steps
+
+    def get_probability_of_sampling_solution(self, key) -> float:
+        return 0.75**self.n
+
+    def get_random_sample(
+        self, rng: np.random.Generator
+    ) -> tuple[np.ndarray, np.ndarray]:
+        return (
+            self.generate_random_assignment(self.n_assignment, rng),
+            self.generate_walk_steps(self.n_walk_steps, rng),
+        )
 
 
 def schoening_solve(
     inst: SatInstance,
     *,
     rng: np.random.Generator,
-    error: Optional[float] = None,
+    error: float,
 ) -> Optional[Assignment]:
     """
-    Find a satisfying assignment of a 3-SAT formula by using Schöning's algorithm,
-    which incrementally flipping bits contained in unsatisfied clauses.
+    Find a satisfying assignment of a 3-SAT formula by using Schoening's algorithm,
+    which incrementally flips assigned variables contained in unsatisfied clauses.
 
     Args:
         inst: The 3-SAT Instance for which to find a satisfying assignment.
         rng: Random number generator.
         error: Allowed failure probability.
-        stats: Object that keeps track of statistics about evaluation queries to the SAT instance.
 
     Returns:
         Satisfying assignment if found, None otherwise.
     """
+    # schoening's algorithm is defined for 3-SAT instances.
     assert inst.k == 3
 
-    # prepare search domain (all randomness used by Schoening's algorithm)
-    n = inst.n
-    assignments = [np.array(x, dtype=int) for x in itertools.product([-1, 1], repeat=n)]
-    steps = [np.array(s, dtype=int) for s in itertools.product([0, 1, 2], repeat=3 * n)]
-    domain = itertools.product(assignments, steps)
-
     # find a choice of randomness that makes Schoening's algorithm accept
-    randomness = search(
-        domain,
-        key=lambda x: schoening_with_randomness(inst, x) is not None,
+    randomness = search_by_sampling_with_replacement(
+        SchoeningDomain(inst.n, inst.n, 3 * inst.n),
+        key=named_oracle("inst.schoening")(
```

Comment:
why is this named? remember, we always want to count number of calls to `schoening_with_randomness`, not the intermediate oracles (as they cannot be compared across variants).

Reply:
My bad, you are right. Will fix.

Reply:
A suggestion: Add to the docs for the benchmanager that you have to pass the function itself into `get_stats`, if you have tagged it with `@oracle`. Similarly you have to call `get_stats` with a string, if you used `@named_oracle(name)`?

As the output of supplying a string instead of the function yields the error that `object {name} has not been benchmarked`, which might be confusing for users?